### PR TITLE
Add system setting table with cache and also add cache supports for user setting

### PR DESCRIPTION
**Pull request by lunny**
- [x] Add a new table `system_setting` to store system setting. And it also has been stored into cache when enabled. 
- [x] Move some related code into `system`
- [x] Also add cache supports for user settings.
- [x] Add a UI for site admin to edit system settings, this will reuse admin configuration page.
- [x] Move some settings from `app.ini` to the system setting table, so then all these setting could be updated immediately with no necessary to reboot Gitea.
    - Move `DisableGravatar` from ini to table `system_setting` as a record with key `disable_gravatar`
    - Move `EnableFederatedAvatar` from ini to table `system_setting` as a record with key `enable_federated_avatar`

image

### CODE REVIEW

**Comment by silverwind** [comment]
Seems like a good idea to have some settings UI-editable but I think gitea must stay fully configurable from ini. Do you plan on writing modified settings back into the ini?

**Comment by lunny** [comment]
> Seems like a good idea to have some settings UI-editable but I think gitea must stay fully configurable from ini. Do you plan on writing modified settings back into the ini?

No. I plan to move most configurations into the database and will not write back to ini and there will be migration to copy user's old setting from ini into database.

Which configuration should be moved into database and anothers will not?
I think it depends on if it needs to be changed when installation. i.e.
Database related configuration should be kept in `ini` file, `HTTP` listen related, `storage` configuration and etc.

And those i.e. pagination numbers, some cron tasks enable/disable.

What's the benefits? Less restarts. 
Currently, even if you change a pagination number you have to restart Gitea. It must be changed and the configuration is unnecessary to be kept in `ini` file.

**Comment by silverwind** [comment]
Migration seems like a must-have, yeah. Maybe even emit warnings on CLI for migrated settings in the ini so users know which ones to remove there.

Cron tasks seems like a good canidate, it always felt odd to me to having to configure those in the ini.

File: models/system/setting.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+// Setting is a key value store of user settings
+type Setting struct {
+	ID           int64  `xorm:"pk autoincr"`
+	SettingKey   string `xorm:"varchar(255) unique"` // ensure key is always lowercase
+	SettingValue string `xorm:"text"`
+}
+
+// TableName sets the table name for the settings struct
+func (s *Setting) TableName() string {
+	return "system_setting"
```
Comment (lafriks) [inline models/system/setting.go:34]:
```suggestion
	return "system_settings"
```

Comment (wxiaoguang) [reply models/system/setting.go:34]: I prefer singular. Most other table names are also singular.

**Comment by mscherer** [comment]
Ini files is easier to manage using a configuration management system (eg, ansible, puppet, etc) and moving them to the database will make it harder to collaborate using git for sysadmins.  

I think a CLI command to change the settings would make sense and avoid inflicting too much pain to sysadmins who wish to use config management, and also maybe one to prevent UI editing (if possible) to prevent errors and manual change.

**Comment by lunny** [comment]
> Ini files is easier to manage using a configuration management system (eg, ansible, puppet, etc) and moving them to the database will make it harder to collaborate using git for sysadmins.
> 
> I think a CLI command to change the settings would make sense and avoid inflicting too much pain to sysadmins who wish to use config management, and also maybe one to prevent UI editing (if possible) to prevent errors and manual change.

This will move some of them which Gitea should not be restarted when changing.

File: modules/system/setting.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	return cache.GetString(genKey(key), func() (string, error) {
+		res, err := system.GetSetting(key)
+		if err != nil {
+			return "", err
+		}
+		return res.SettingValue, nil
+	})
+}
+
+func GetSetingBool(key string) bool {
```
Comment (mscherer) [inline modules/system/setting.go]: I guess this one should be GetSettingBool for consistency.

File: web_src/less/themes/theme-arc-green.css [hunk 3]
```diff
@@ -0,0 +1,762 @@
+.chroma .hl {
+  background-color: #3f424d;
+}
+/* LineHighlight */
+.chroma .lnt {
```
Comment (6543) [inline web_src/less/themes/theme-arc-green.css]: why we have this css in less now?

Comment (lunny) [reply web_src/less/themes/theme-arc-green.css]: Good Catch. done.

File: models/migrations/v216.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
```
Comment (6543) [inline models/migrations/v216.go]:
```suggestion
// Copyright 2022 The Gitea Authors. All rights reserved.
```

**Review by 6543 — APPROVED** [review]

File: web_src/js/features/admin/config.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+      type: 'POST',
+      data: {
+        _csrf: csrfToken,
+        key: $this.attr('name'),
+        value: $this.is(':checked'),
+        version: $this.attr('version'),
+      }
+    }).done((resp) => {
+      if (resp) {
+        console.info(resp);
```
Comment (silverwind) [inline web_src/js/features/admin/config.js]:
```suggestion
```

Don't litter the console.

File: routers/web/admin/config.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+	ctx.Data["CacheInterval"] = setting.CacheService.Interval
+
+	ctx.Data["CacheConn"] = shadowPassword(setting.CacheService.Adapter, setting.CacheService.Conn)
+	ctx.Data["CacheItemTTL"] = setting.CacheService.TTL
+
+	sessionCfg := setting.SessionConfig
+	if sessionCfg.Provider == "VirtualSession" {
+		var realSession session.Options
+		if err := json.Unmarshal([]byte(sessionCfg.ProviderConfig), &realSession); err != nil {
+			log.Error("Unable to unmarshall session config for virtualed provider config: %s\nError: %v", sessionCfg.ProviderConfig, err)
```
Comment (silverwind) [inline routers/web/admin/config.go]:
```suggestion
			log.Error("Unable to unmarshall session config for virtual provider config: %s\nError: %v", sessionCfg.ProviderConfig, err)
```

"virtualed" is certainly not a english word :)

**Comment by silverwind** [comment]
Needs rebase because of `v219.go`

**Comment by lunny** [comment]
> Needs rebase because of `v219.go`

Thanks, done.

File: web_src/js/features/admin/config.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import $ from 'jquery';
+
+export function initAdminConfigs() {
+  const isAdminConfigPage = window.config.pageData.adminConfigPage;
+  if (!isAdminConfigPage) return;
+
+  const {appSubUrl, csrfToken} = window.config;
```
Comment (silverwind) [inline web_src/js/features/admin/config.js]:
```suggestion

const {appSubUrl, csrfToken, pageData} = window.config;

export function initAdminConfigs() {
  const isAdminConfigPage = pageData?.adminConfigPage;
  if (!isAdminConfigPage) return;
```

**Comment by silverwind** [comment]
Gave it a try, changing any setting currently fails for me:

```
2022/08/16 20:12:23 .../web/admin/config.go:209:ChangeConfig() [E] [62fbde07] set setting failed: System setting[disable_gravatar] has been updated by another thread
2022/08/16 20:12:28 .../web/admin/config.go:209:ChangeConfig() [E] [62fbde0b-2] set setting failed: System setting[enable_federated_avatar] has been updated by another thread
```

The UI does not show these errors, and I think we really need a toast-style error popup in this case. At least we should revert these toggle boxes to their previous state on error, popup can be added separately.

**Comment by lunny** [comment]
> 

All done.

**Comment by silverwind** [comment]
Works now. I'm still not convinced of the value of singling out these two settings, or whether this versioning that seems to be going on when changing these settings is necessary.

Care to explain why the versioning is added? Is it so one can roll back to previous configs?

**Review by silverwind — APPROVED** [review]
I guess we can see how it goes.

File: routers/install/install.go [hunk 8]
```diff
@@ -147,8 +149,19 @@ func Install(ctx *context.Context) {
...
-	form.DisableGravatar = setting.DisableGravatar
-	form.EnableFederatedAvatar = setting.EnableFederatedAvatar
+	disableGravatarSetting, _ := system_model.GetSetting(system_model.KeyPictureDisableGravatar)
+	if disableGravatarSetting != nil {
+		form.DisableGravatar = disableGravatarSetting.GetValueBool()
+	} else {
+		form.DisableGravatar = false
+	}
+
+	enableFederatedAvatarSetting, _ := system_model.GetSetting("")
```
Comment (6543) [inline routers/install/install.go]: key missing ?

Comment (6543) [reply routers/install/install.go]: @lunny shouldn't this contain "picture.enable_federated_avatar" ?

*KeyPictureEnableFederatedAvatar

Comment (lunny) [reply routers/install/install.go]: Good catch, fixed.

File: modules/system/setting.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	return cache.GetString(genKey(key), func() (string, error) {
+		res, err := system.GetSetting(key)
+		if err != nil {
+			return "", err
+		}
+		return res.SettingValue, nil
+	})
+}
+
+func GetSettingBool(key string) bool {
```
Comment (6543) [inline modules/system/setting.go:31]:
```suggestion
// GetSettingBool return bool value of setting,
// none existing keys and errors are ignored and result in false
func GetSettingBool(key string) bool {
```

**Review by 6543 — APPROVED** [review]

**Comment by yardenshoham** [comment]
I think this breaks gitpod
image

```log
        /home/gitpod/go/src/runtime/panic.go:260 (0x459475)
                panicmem: panic(memoryError)
        /home/gitpod/go/src/runtime/signal_unix.go:835 (0x459445)
                sigpanic: panicmem()
        /workspace/gitea/models/db/context.go:78 (0xb9ca65)
                GetEngine: enginedInterface := ctx.Value(enginedContextKey)
        /workspace/gitea/models/system/setting.go:96 (0xcaa93e)
                GetSettings: if err := db.GetEngine(db.DefaultContext).
        /workspace/gitea/models/system/setting.go:80 (0xcaa765)
                GetSetting: v, err := GetSettings([]string{key})
        /workspace/gitea/routers/install/install.go:152 (0x206feba)
                Install: disableGravatarSetting, _ := system_model.GetSetting(system_model.KeyPictureDisableGravatar)
        /workspace/gitea/modules/web/wrap_convert.go:47 (0x1db4f13)
                convertHandler.func3: t(ctx)
        /workspace/gitea/modules/web/wrap.go:41 (0x1db33e4)
                wrapInternal.func1: done, deferrable := handler(resp, req, others...)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 (0x19b1dd5)
                (*Mux).routeHTTP: h.ServeHTTP(w, r)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/gitea/routers/install/install.go:89 (0x206f9ea)
                Init.func1.1: next.ServeHTTP(resp, ctx.Req)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/gitea/routers/install/routes.go:80 (0x2075071)
                installRecovery.func1.1: next.ServeHTTP(w, req)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/go/pkg/mod/gitea.com/go-chi/session@v0.0.0-20211218221615-e3605d8b28b8/session.go:257 (0x11a77aa)
                Sessioner.func1.1: next.ServeHTTP(w, req)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/gitea/modules/web/wrap.go:110 (0x1db41e8)
                WrapWithPrefix.func1.1: next.ServeHTTP(resp, req)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/gitea/routers/common/middleware.go:79 (0x1e7a4e2)
                Middlewares.func2.1: next.ServeHTTP(resp, req)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/gitea/modules/web/routing/logger_manager.go:123 (0x1daf0f3)
                (*requestRecordsManager).handler.func1: next.ServeHTTP(w, req)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/middleware/strip.go:30 (0x1e77d78)
                StripSlashes.func1: next.ServeHTTP(w, r)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/go/pkg/mod/github.com/chi-middleware/proxy@v1.1.1/middleware.go:37 (0x1e74516)
                ForwardedHeaders.func1.1: h.ServeHTTP(w, r)
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/gitea/routers/common/middleware.go:32 (0x1e7a331)
                Middlewares.func1.1: next.ServeHTTP(context.NewResponse(resp), req.WithContext(ctx))
        /home/gitpod/go/src/net/http/server.go:2109 (0x94f18e)
                HandlerFunc.ServeHTTP: f(w, r)
        /workspace/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:88 (0x19afb4f)
                (*Mux).ServeHTTP: mx.handler.ServeHTTP(w, r)
        /workspace/gitea/modules/web/route.go:200 (0x1db280d)
                (*Route).ServeHTTP: r.R.ServeHTTP(w, req)
        /home/gitpod/go/src/net/http/server.go:2947 (0x95258b)
                serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
        /home/gitpod/go/src/net/http/server.go:1991 (0x94dc66)
                (*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
        /home/gitpod/go/src/runtime/asm_amd64.s:1594 (0x477140)
                goexit: BYTE    $0x90   // NOP
```

cc @anbraten

**Comment by techknowlogick** [comment]
@wxiaoguang I had updated try.gitea.io to use local avatars to test the switch, can switch it back.

**Comment by SuperSandro2000** [comment]
>  Move some settings from app.ini to the system setting table, so then all these setting could be updated immediately with no necessary to reboot Gitea. 

Is this really necessary? With that it is no longer possible to easily manage the settings decoratively.

**Comment by lunny** [comment]
> > Move some settings from app.ini to the system setting table, so then all these setting could be updated immediately with no necessary to reboot Gitea.
> 
> Is this really necessary? With that it is no longer possible to easily manage the settings decoratively.

We can provide a command to help set/get them later. Some users don't want to restart Gitea even if he just change a pagination number.

**Comment by SuperSandro2000** [comment]
> We can provide a command to help set/get them later. Some users don't want to restart Gitea even if he just change a pagination number.

For NixOS it would probably be the easiest if the value just appeared again in the config file. No need to develop something new.

**Comment by kdumontnu** [comment]
Should this be labeled as a breaking change? I know it's a little late, but aren't we essentially ~deprecating~ breaking support for some ini settings?

I suggest we think through and document how config settings that are being migrated from ini to the database will be supported in the future and what the interaction between database and ini is (for example, does ini override db setting or vice-versa?). I'm a little confused how server settings should be set and maintained now, as there are now two places to configure server settings (three if you include some support for env variable overrides).

**Comment by rmbleeker** [comment]
I haven't seen a satisfying answer yet to a point that's been brought up a few times already, which is: how would a sysadmin (like myself) set the options that now reside in the database during an automated installation of Gitea? I personally use Ansible for my automation needs, I have a Jinja2 template of the ini file and all the options for Gitea in yaml. This is very easy to maintain and makes for a quick installation/update, while calling gitea from the command line to change options in a database would not be. Having to use the GUI post-installation to properly configure your software completely defeats the purpose of automation.

**Pull request merged**